src/khepri_machine.erl

```diff
@@ -2322,3 +2340,126 @@ convert_state(State, 0, 1) ->
+%% The changes glossed over by the snapshot may include projection
+%% registrations and unregistrations so we need to initialize new projections
+%% and delete unregistered ones, and we need to ensure that the projection
+%% tables are up to date for any projections which didn't change.
+%%
+%% @private
+
+update_projections(OldState, NewState) ->
+    OldTree = get_tree(OldState),
+    OldProjections = set_of_projections(get_projections(OldState)),
+    NewTree = get_tree(NewState),
+    NewProjections = set_of_projections(get_projections(NewState)),
+
+    CommonProjections = sets:intersection(OldProjections, NewProjections),
+    DeletedProjections = sets:subtract(OldProjections, CommonProjections),
+    CreatedProjections = sets:subtract(NewProjections, CommonProjections),
+
+    %% Tear down any projections which were unregistered.
+    lists:foreach(
+      fun({_Pattern, Projection}) ->
+              _ = khepri_projection:delete(Projection)
+      end, sets:to_list(DeletedProjections)),
+
+    %% Initialize any new projections which were registered.
+    lists:foreach(
+      fun({Pattern, Projection}) ->
+              ok = restore_projection(Projection, NewTree, Pattern)
+      end, sets:to_list(CreatedProjections)),
+
+    %% Update in-place any projections which were not changed themselves (i.e.
+    %% the projection name, function and pattern) between old and new states.
+    %% To do this we will find the matching nodes in the old and new tree for
+    %% the projection's pattern and trigger the projection based on each
+    %% matching path's old and new properties.
+    lists:foreach(
+      fun({Pattern, Projection}) ->
+              ok = update_projection(Pattern, Projection, OldTree, NewTree)
+      end, sets:to_list(CommonProjections)),
+
+    ok.
+
+-spec set_of_projections(ProjectionTree) -> Projections when
+      ProjectionTree :: khepri_machine:projection_tree(),
+      Element :: {khepri_path:native_pattern(),
+                     khepri_projection:projection()},
+      Projections :: sets:set(Element).
+%% Folds the set of projections in a projection tree into a version 2 {@link
+%% sets:set()}.
+%%
+%% @private
+
+set_of_projections(ProjectionTree) ->
+    khepri_pattern_tree:fold(
+      ProjectionTree,
+      fun(Pattern, Projections, Acc) ->
+              lists:foldl(
+                fun(Projection, Acc1) ->
+                        Entry = {Pattern, Projection},
+                        sets:add_element(Entry, Acc1)
+                end, Acc, Projections)
+      end, sets:new([{version, 2}])).
+
+update_projection(Pattern, Projection, OldTree, NewTree) ->
+    TreeOptions = #{props_to_return => ?PROJECTION_PROPS_TO_RETURN,
+                    include_root_props => true},
+    case khepri_tree:find_matching_nodes(OldTree, Pattern, TreeOptions) of
+        {ok, OldMatchingNodes} ->
+            case khepri_tree:find_matching_nodes(NewTree, Pattern, TreeOptions) of
```

Comment:
This line is too long. Could you please assign the result of `khepri_tree:find_matching_nodes(NewTree, Pattern, TreeOptions)` to a variable and use the variable in the `case`?